pretty/marshal.go

```diff
@@ -0,0 +1,232 @@
+			st.writeIndent()
+			b := strconv.AppendInt(st.scratch[:0], int64(i), 10)
+			st.writeBytes(b)
+			st.write(": ")
+			if err := st.marshal(v); err != nil {
+				return err
+			}
+			st.writeBytes(st.cfg.Sep)
+			st.writeBytes(st.cfg.Line)
+			if err := st.checkErr(); err != nil {
+				return err
+			}
+		}
+		st.count = st.count[0 : len(st.count)-1]
+		st.writeIndent()
+		st.write("]")
+		return st.checkErr()
+	case ipld.ReprKind_Bool:
+		v, _ := n.AsBool()
+		switch v {
+		case true:
+			st.write("bool node: true")
+			return st.checkErr()
+		case false:
+			st.write("bool node: false")
+			return st.checkErr()
+		default:
+			panic("unreachable")
+		}
+	case ipld.ReprKind_Int:
+		v, _ := n.AsInt()
+		st.write("int node: ")
+		b := strconv.AppendInt(st.scratch[:0], int64(v), 10)
+		st.writeBytes(b)
+		return st.checkErr()
+	case ipld.ReprKind_Float:
+		v, _ := n.AsFloat()
+		st.write("float node: ")
+		b := strconv.AppendFloat(st.scratch[:0], v, 'g', -1, 64)
+		st.writeBytes(b)
+		return st.checkErr()
+	case ipld.ReprKind_String:
+		v, _ := n.AsString()
+		st.write("string node: ")
+		return codectools.WriteQuotedEscapedString(v, st.w)
+	case ipld.ReprKind_Bytes: // we're going to use hexdump for this, which is pretty wild.  Not meant for parsing!
+		v, _ := n.AsBytes()
+		st.write("bytes node:")
+		st.count = append(st.count, 0)
+		for _, s := range strings.Split(hex.Dump(v), "\n") {
+			if s == "" {
+				continue
+			}
+			st.writeBytes(st.cfg.Line) // this will not be particularly readable if Line is nothing.
+			st.writeIndent()
+			st.writeBytes([]byte{'|', ' '})
+			st.write(s)
+		}
+		st.count = st.count[0 : len(st.count)-1]
+		return st.checkErr()
+	case ipld.ReprKind_Link:
+		v, _ := n.AsLink()
+		st.write("link node: ")
+		st.write(v.String())
+		return st.checkErr()
+	case ipld.ReprKind_Invalid:
+		panic("invalid node encountered")
+	default:
+		panic("unreachable")
+	}
+}
+
+// emitFlat is used for map keys.
+// Mostly map keys are strings... but they can be structs, too, if we're dealing with schemas and they have a string representation.
+// We don't want to do a full marshal recursively for this for two reasons:
+//  - if it's not a string, with indentation it'd be a mess;
+//  - we don't to repeat the type info for every key in a map because it's too bulky and not interesting.
+func (st *encodeState) emitFlat(n ipld.Node) error {
+	switch n.ReprKind() {
+	case ipld.ReprKind_String:
+		s, _ := n.AsString()
+		return codectools.WriteQuotedEscapedString(s, st.w)
+	case ipld.ReprKind_Map:
+		panic("nyi")
+	}
+	panic("emitFlat: unhandled node kind")
+}
+
+func (st *encodeState) writeIndent() {
+	for i := 0; i < len(st.count); i++ {
+		st.writeBytes(st.cfg.Indent)
+	}
+}
+
+func (st *encodeState) write(s string) {
+	if st.err == nil {
+		_, st.err = st.w.Write([]byte(s))
+	}
+}
+func (st *encodeState) writeBytes(b []byte) {
+	if st.err == nil {
+		_, st.err = st.w.Write(b)
+	}
+}
+func (st *encodeState) checkErr() error {
```

Comment:
I'm not sure I get the point of this func

Reply:
You're right, I should remove this.  I must've expected it would do more work at first, and then didn't "gc" when it didn't turn out to be so.